File: q2/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* addTwo(struct ListNode* l1, struct ListNode* l2, int carryOn);
struct ListNode* addTwoNum(struct ListNode* l1, struct ListNode* l2, int carryOn);

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    return addTwoNum(l1, l2, 0);
}

/* recursive function that performs addition of two singly linked lists */
struct ListNode* addTwoNum(struct ListNode* l1, struct ListNode* l2, int carryOn) {

    struct ListNode* carry = NULL;
    if (carryOn) {
        carry = malloc(sizeof(struct ListNode));
        carry -> val = 1; 
        carry -> next = NULL;
    }

    // base cases
    if (!l1 && !l2) return carry;
    if (!l1) return addTwoNum(l2, carry, 0);
    if (!l2) return addTwoNum(l1, carry, 0); 

    // addition performed here
    struct ListNode* added = addTwo(l1, l2, carryOn);
    added -> next = addTwoNum(l1 -> next, l2 -> next, l1 -> val + l2 -> val + carryOn > 9);

    return added; 
}

/* function that performs the addition of two single nodes */
struct ListNode* addTwo(struct ListNode* l1, struct ListNode* l2, int carryOn) {

    struct ListNode* added = malloc(sizeof(struct ListNode));
    added -> val = (l1 -> val + l2 -> val + carryOn) % 10;
    added -> next = NULL; 

    return added;
}

File: q2/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        head = ListNode()
        tail = head
        carry = 0

        # ends when there are no values to be added or carry on values 
        while l1 is not None or l2 is not None or carry != 0:

            digit1 = l1.val if l1 is not None else 0
            digit2 = l2.val if l2 is not None else 0
            sm = digit1 + digit2 + carry

            # append the added result of the digits to solution
            added = ListNode(sm % 10)
            tail.next = added
            tail = tail.next 

            # reset values for next iteration 
            carry = sm // 10
            l1 = l1.next if l1 is not None else None
            l2 = l2.next if l2 is not None else None 

        result = head.next
        head.next = None
        return result 


    